assets/js/amp-editor-blocks.js

```diff
@@ -0,0 +1,374 @@
+				if ( 'container' === option.value || 'intrinsic' === option.value ) {
+					return true;
+				}
+			} else if (
+				'core-embed/vimeo' === blockName ||
+				'core-embed/dailymotion' === blockName ||
+				'core-embed/hulu' === blockName ||
+				'core-embed/reddit' === blockName
+			) {
+				if ( 'container' === option.value || 'intrinsic' === option.value || 'nodisplay' === option.value ) {
+					return true;
+				}
+			} else if ( 'core-embed/soundcloud' === blockName ) {
+				if ( 'fixed-height' !== option.value ) {
+					return true;
+				}
+			}
+
+			layoutOptions.push( { value: option.value, label: option.label } );
+		} );
+
+		return layoutOptions;
+	};
+
+	/**
+	 * Add extra data-amp-layout attribute to save to DB.
+	 *
+	 * @param {Object} props Properties.
+	 * @param {string} blockType Block type.
+	 * @param {Object} attributes Attributes.
+	 * @return {*} Props.
+	 */
+	component.addAMPExtraProps = function addAMPExtraProps( props, blockType, attributes ) {
+		var ampAttributes = {};
+		if ( ! attributes.ampLayout && ! attributes.ampNoLoading ) {
+			return props;
+		}
+
+		if ( attributes.ampLayout ) {
+			ampAttributes[ 'data-amp-layout' ] = attributes.ampLayout;
+		}
+		if ( attributes.ampNoLoading ) {
+			ampAttributes[ 'data-amp-noloading' ] = attributes.ampNoLoading;
+		}
+
+		return _.assign( ampAttributes, props );
+	};
+
+	/**
+	 * Add AMP attributes (in this test case just ampLayout) to every core block.
+	 *
+	 * @param {Object} settings Settings.
+	 * @param {string} name Block name.
+	 * @return {*} Settings.
+	 */
+	component.addAMPAttributes = function addAMPAttributes( settings, name ) {
+		var mediaBlocks = [
+			'core/image',
+			'core/video',
+			'core/audio'
+		];
+
+		// Gallery settings for shortcode.
+		if ( 'core/shortcode' === name ) {
+			if ( ! settings.attributes ) {
+				settings.attributes = {};
+			}
+			settings.attributes.ampCarousel = {
+				type: 'boolean'
+			};
+		}
+
+		// Layout settings for embeds and media blocks.
+		if ( 0 === name.indexOf( 'core-embed' ) || -1 !== mediaBlocks.indexOf( name ) ) {
+			if ( ! settings.attributes ) {
+				settings.attributes = {};
+			}
+			settings.attributes.ampLayout = {
+				type: 'string'
+			};
+			settings.attributes.ampNoLoading = {
+				type: 'boolean'
+			};
+		}
+		return settings;
+	};
+
+	/**
+	 * Filters blocks edit function of all blocks.
+	 *
+	 * @param {Function} BlockEdit Edit function.
+	 * @return {Function} Edit function.
+	 */
+	component.filterBlocksEdit = function filterBlocksEdit( BlockEdit ) {
+		var el = wp.element.createElement;
+
+		return function( props ) {
+			var attributes = props.attributes,
+				isSelected = props.isSelected,
+				name = props.name,
+				ampLayout,
+				inspectorControls,
+				width;
+
+			ampLayout = attributes.ampLayout;
+
+			if ( 'core/shortcode' === name ) {
+				// Lets remove amp-carousel from from edit view.
+				if ( component.hasGalleryShortcodeCarouselAttribute( attributes.text ) ) {
```

Comment:
I'm getting a JS error here when first adding a shortcode because `text` prop is undefined. So I think the value passed into this function should be `attributes.text || ''`.